pkg/route/secretmanager/manager_test.go

```diff
@@ -0,0 +1,240 @@
+package secretmanager
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/openshift/library-go/pkg/route/secret"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+type routeSecret struct {
+	routeName  string
+	secretName string
+}
+
+type fakeSecretMonitor struct {
+	err error
+}
+
+func (sm *fakeSecretMonitor) AddSecretEventHandler(_ context.Context, _ string, _ string, _ cache.ResourceEventHandler) (secret.SecretEventHandlerRegistration, error) {
+	return nil, sm.err
+}
+func (sm *fakeSecretMonitor) RemoveSecretEventHandler(_ secret.SecretEventHandlerRegistration) error {
+	return sm.err
+}
+func (sm *fakeSecretMonitor) GetSecret(_ secret.SecretEventHandlerRegistration) (*corev1.Secret, error) {
+	return nil, sm.err
+}
+
+func TestRegisterRoute(t *testing.T) {
+	namespace := "ns"
+
+	scenarios := []struct {
+		name               string
+		rs                 []routeSecret
+		sm                 fakeSecretMonitor
+		expectHandlersKeys []string
+		expectErr          int
+	}{
+		{
+			name: "route can be registered only once with any secret",
+			rs: []routeSecret{
+				{routeName: "route", secretName: "secret"},
+			},
+			expectHandlersKeys: []string{namespace + "/route"},
+			expectErr:          0,
+		},
+		{
+			name: "same route cannot be registered again with same secret",
+			rs: []routeSecret{
+				{routeName: "route1", secretName: "secret1"},
+				{routeName: "route1", secretName: "secret1"},
+			},
+			expectHandlersKeys: []string{namespace + "/route1"},
+			expectErr:          1,
+		},
+		{
+			name: "same route cannot be registered again with different secrets",
+			rs: []routeSecret{
+				{routeName: "route1", secretName: "secret1"},
+				{routeName: "route1", secretName: "secret2"},
+				{routeName: "route1", secretName: "secret3"},
+			},
+			expectHandlersKeys: []string{namespace + "/route1"},
+			expectErr:          2,
+		},
+		{
+			name: "different routes can be registered with same secret",
+			rs: []routeSecret{
+				{routeName: "route1", secretName: "secret1"},
+				{routeName: "route2", secretName: "secret1"},
+			},
+			expectHandlersKeys: []string{namespace + "/route1", namespace + "/route2"},
+			expectErr:          0,
+		},
+		{
+			name: "different routes can be registered with different secrets",
+			rs: []routeSecret{
+				{routeName: "route1", secretName: "secret1"},
+				{routeName: "route2", secretName: "secret2"},
+			},
+			expectHandlersKeys: []string{namespace + "/route1", namespace + "/route2"},
+			expectErr:          0,
+		},
+		{
+			name: "error while adding SecretEventHandler",
+			rs: []routeSecret{
+				{routeName: "route", secretName: "secret"},
+			},
+			sm:        fakeSecretMonitor{err: fmt.Errorf("some error")},
+			expectErr: 1,
+		},
+	}
+	for _, s := range scenarios {
+		t.Run(s.name, func(t *testing.T) {
+			mgr := NewManager(nil, nil).WithSecretMonitor(&s.sm)
```

Comment:
We need this function to update the secret monitor during runtime, even after initializing with `NewManager`.  For easier update, we need to keep it.

Reply:
For now, I'm using it in unit test cases only, to update the secretMonitor.

https://github.com/openshift/library-go/pull/1626/files#diff-e241757c6b956d7bfe3536af68c35f09c0d562a4332d54cd0b03dfafd9e46a9bR161-R173

Idea is to avoid any errors from `RegisterRoute()` with blank `fake.SecretMonitor{}` , and then change it to input (test case) secretMonitor, before calling `UnregisterRoute` and `GetSecret`.

I'm not yet 100% sure whether this will be required by the consumer (o/router), if not we can refactor it later.

Reply:
Update: I think it's unlikely that `o/router` will make use of  `WithSecretMonitor` because it will not change the monitor after initializing `Manager`.

I've refactored the tests and also removed `WithSecretMonitor` in https://github.com/openshift/library-go/pull/1626/commits/a417674c400d5866bd59fdfad0f84458b53c8958